spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/DispatchingStreamListenerMessageHandler.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binding;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * An {@link AbstractReplyProducingMessageHandler} that delegates
+ * @author Marius Bogoevici
+ */
+final class DispatchingStreamListenerMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final Collection<ConditionalStreamListenerHandler> handlerMethods;
+
+	private final EvaluationContext evaluationContext;
+
+	DispatchingStreamListenerMessageHandler(Collection<ConditionalStreamListenerHandler> handlerMethods,
+			EvaluationContext evaluationContext) {
+		Assert.notEmpty(handlerMethods, "'handlerMethods' cannot be empty");
+		Assert.notNull(evaluationContext, "'evaluationContext' cannot be empty");
+		this.handlerMethods = handlerMethods;
+		this.evaluationContext = evaluationContext;
+	}
+
+	@Override
+	protected boolean shouldCopyRequestHeaders() {
+		return false;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Collection<ConditionalStreamListenerHandler> matchingHandlers = findMatchingHandlers(requestMessage);
+		if (matchingHandlers.size() == 0) {
+			if (logger.isWarnEnabled()) {
+				logger.warn("Cannot find a @StreamListener matching " + requestMessage);
+			}
+			return null;
+		}
+		else if (matchingHandlers.size() > 1) {
+			for (ConditionalStreamListenerHandler matchingHandler : matchingHandlers) {
+				if (!matchingHandler.isVoid()) {
+					throw new MessagingException(
+							"Multiple matching methods cannot return values for " + requestMessage);
+				}
+			}
+			for (ConditionalStreamListenerHandler matchingMethod : matchingHandlers) {
```

Comment:
chatted a bit with @artembilan about this. One other possibility is to nix conditions on value-returning methods altogether - it eliminates the potential of erratic behaviour as to the return values of the methods. This would turn this feature into a SpEL-matching event handler that would be suitable for Event Streaming/Event Sourcing scenario and would avoid the risk of rebuilding a router via `@StreamListener`. The use case @dturanski has brought up definitely doesn't require return values . @garyrussell , wdyt?